Meta class timing model

luojing1211 opened the pull request:
This is a metaclass TimingModel update for solving the problem mentioned at issue #218

luojing1211 commented:
All TimingModel classes are added to __init__.py 
The class member 'name' is changed to 'is_regisiter'. If 'is_register' true, the class will be registered to the model list. The name will be parsed to the list automatically.

cdeil commented on pint/models/timing_model.py:
```diff
@@ -96,6 +97,21 @@ def use_cached_results(*args, **kwargs):
+class ModelMeta(abc.ABCMeta):
+    """
+    This is a Meta class for timing model registeration. In order ot get a
+    timing model registered, a member called 'name' has to be provided in a
```
The docstring here still says "name", but it's now "is_register" I think.

"is_register=True" sounds a little off to me, because it's a mix between "is_registered" and "register me". Maybe call it "register"?

luojing1211 commented:
Quick question:
    How do we swith on Python3 test for travis?

cdeil commented:
> How do we swith on Python3 test for travis?

We have to do a significant extension of `.travis.yml` to specify multiple builds.
There's different ways to specify the build matrix.
Here's a simple example: https://github.com/rootpy/rootpy/blob/master/.travis.yml
Here's a more complicated example: https://github.com/gammapy/gammapy/blob/master/.travis.yml

It usually takes a few hours and several commits to get `.travis.yml` right, especially if you haven't done it before.  So I don't think you should attempt it in this PR. I can make a new PR extending `.travis.yml` in the coming days.

jellis18 commented:
I've started testing this with `python3`. There are lots of little fixes that are needed but as @cdeil says, the whole point was to make `PINT` `python3` compatible so I think this may be the place. although maybe we should update the title of the PR.

luojing1211 commented:
I think we should merge this branch, since this is on optimization on the PINT structure. We should make a new pull request for fixing compatibility of Python 3.x

jellis18 commented:
@luojing1211, ok but what is this PR supposed to fix then if not the Python 3 compatibility?

luojing1211 commented:
This is to change the way of timing model selection. It is better than the
current method.

On Wed, Feb 22, 2017 at 3:39 PM, Justin Ellis <notifications@github.com>
wrote:

> @luojing1211 <https://github.com/luojing1211>, ok but what is this PR
> supposed to fix then if not the Python 3 compatibility?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nanograv/PINT/pull/220#issuecomment-281812662>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFFAg1o6KwW_QqaP1pIoUOk_4fxyygEHks5rfKr6gaJpZM4Lrl0f>
> .
>

luojing1211 commented:
Hi @ALL, 

I added some new features for this branch:
* All timing model components has metaclass for registering the components. 
* Add very basic components information class and order structure. 
    PR #170 seems over complicated. 
* Modified par file print out function. It handles basic parameters grouping and sorting. 
   This feature will fix issue #78   
Problem:
* No unittest for modified as_parfile() function . 
If there is no good way to test it yet. I would recommend this branch get merged.

paulray commented:
As a unit test for as_parfile(), how about writing a moderately complicated model constructed programmatically (not from a par file), writing it out as a par file, then reading it back in, and verify that the phases computed (for a few random TOAs) are the same between the original model and the one read in.

luojing1211 commented:
This is a good idea. Besides, I would like to add a convenient function for getting the information  from the toas class. Which module do you think this function should be in? Maybe model builder is a good place?

paulray commented:
@jellis18 I see that the coverage/coveralls check is failing.  Is it simply a test to see if the coverage decreases and that constitutes a fail, or is it more subtle?

paulray commented:
> Besides, I would like to add a convenient function for getting the information from the toas class. Which module do you think this function should be in? Maybe model builder is a good place?

I don't understand what you are asking here. Can you clarify @luojing1211 ?

luojing1211 commented:
Sorry for the confusion. I was talking about the convenient function for
getting the "EPHEM", "CLK", information from a parfile.

On Wed, Mar 15, 2017 at 1:58 PM, Paul Ray <notifications@github.com> wrote:

> Besides, I would like to add a convenient function for getting the
> information from the toas class. Which module do you think this function
> should be in? Maybe model builder is a good place?
>
> I don't understand what you are asking here. Can you clarify @luojing1211
> <https://github.com/luojing1211> ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nanograv/PINT/pull/220#issuecomment-286845365>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFFAg-S8nQxAx6DRid1qERs-8Q__Vxsaks5rmDTzgaJpZM4Lrl0f>
> .
>

jellis18 commented:
@paulray, I think that is it. There are setting to set the threshold for how much coverage can decrease before failure although I can't find what the defaults are (maybe it is 0.5%). I changed it to 10% and am re-running travis to see if it passes.

I guess it is up to us to set this number though. It may be a good idea to make this stringent so that it requires that any new features have tests. This also brings me back to the protected branch idea. Right now we could merge this even though the tests failed. If we make `PINT/master` a protected branch then the tests have to pass before we can merge.

paulray commented:
@jellis18 thanks for the explanation! For at least a little longer, while PINT has such a small set of developers and is under such active development, I don't think I'd make a protected branch (especially if a 0.5% drop in coverage causes a fail). But, putting in a milestone to do that at some point seems like a good idea. Having travis has been great in keeping us from merging things that break the tests. We definitely need to focus more on building good tests as well and the coverage stuff does help encourage that.

On the other hand, the code quality checker has some insanely strict rules, like 80 column lines and too many local variables. I hope it has some way to tune what it reports. For example, flags me every time I use log.info("message {0}".format(val)), which I use ALL the time.  I don't agree that it is poor style, and I don't really understand how to write it in a different way.

paulray commented:
Is it OK to merge this? I'd like to get at least the parfile writing changes.

luojing1211 commented:
Not yet. This afternoon should be ok.


On Fri, Mar 17, 2017 at 10:02 AM Paul Ray <notifications@github.com> wrote:

> Is it OK to merge this? I'd like to get at least the parfile writing
> changes.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nanograv/PINT/pull/220#issuecomment-287378445>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFFAg9gXOTY0hqx3nZwVEKUBpLsXPsfbks5rmqBpgaJpZM4Lrl0f>
> .
>

luojing1211 commented:
If the check is successful, this branch should be ok to merge.

luojing1211 commented:
Hi @paulray , do you want to double check this branch and merge this one?
